File: Python/medium/24. Swap Nodes in Pairs.py
"""
Given a linked list, swap every two adjacent nodes and return its head.

You may not modify the values in the list's nodes, only nodes itself may be changed.

 

Example:

Given 1->2->3->4, you should return the list as 2->1->4->3.
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return
        dummy = ListNode(0)
        dummy.next = head
        slow = dummy
        fast = slow.next
        while fast and fast.next:
            slow.next = fast.next
            fast.next = slow.next.next
            slow.next.next = fast
            slow = fast
            fast = slow.next
        return dummy.next

File: Java/Medium/24. Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode dummy = new ListNode(0, null);
        ListNode prevNode = dummy;
        ListNode middleNode = head;
        ListNode nextNode = head.next;
        while (middleNode != null && nextNode != null) {
            middleNode.next = nextNode.next;
            nextNode.next = middleNode;
            prevNode.next = nextNode;
            
            if (middleNode.next == null) break;
            prevNode = middleNode;
            middleNode = middleNode.next;
            nextNode = middleNode.next;
        }
        return dummy.next;
    }
}